doc/adrs/155_make_it_possible_to_display_the_semantic_elements_as_a_tree_in_the_selection_dialog.adoc

```diff
@@ -0,0 +1,162 @@
+= [ADR-155] Make it possible to display the semantic elements as a tree in the Selection Dialog
+
+== Context
+
+We have restored the previous deactivated Selection Dialog in the context of the ADR: _[ADR-153] Restore the Selection Dialog support in diagrams
+_.
+
+We want now to improve this Selection Dialog.
+
+The Selection dialog displayed the candidates elements (returned by the `candidatesExpression` in the `SelectionDialogDescription`) as a flat list without any information about the hierarchical structure.
+
+Since the selection dialog is intended to be used to select semantic elements, it will be necessary to have the ability to display them as they are in their structure (as a tree).
+
+image:images/155/currentSelectionDialog.png[Current Selection Dialog, 50%]
+
+=== Current behavior
+
+The Selection dialog opens a subscription to the selection representation. The `SelectionRefreshedEventPayload` returns the Selection that only owns a list of objects to display.
+
+Here is the current GraphQL API:
+
+
+```
+type SelectionRefreshedEventPayload {
+  id: ID!
+  selection: Selection!
+}
+
+type Selection implements Representation {
+  id: ID!
+  metadata: RepresentationMetadata!
+  targetObjectId: String!
+  message: String
+  objects: [SelectionObject!]!
+}
+
+type SelectionObject {
+  id: ID!
+  label: String!
+  iconURL: [String!]!
+}
+```
+
+== Decision
+
+=== Preferred option
+
+==== View model 
+
+The specifier will choose whether the objects returned by the candidates expression will be displayed as a tree or as a flat list. 
+
+The `SelectionDialogDescription` from the Diagram View model will have an additional boolean attributes:
+
+ * `displayedAsTree`: to indicate whether the dialog layout should be a tree.
+
+
+==== Backend and GraphQL API
+
+If the displayedAsTree is true, we will rely on the TreeRepresentation instead of the current SelectionRepresentation.
+
+We will need to contribute a new `IEditingContextRepresentationDescriptionProvider` as the `ModelBrowsersDescriptionProvider` for the Reference widget.
+
+This `DescriptionProvider` will rely on the `SelectionDialogDescription` to compute the `TreeDescription`. 
+
+The GraphQL API will evolve to handle the layout information:
+
+```
+type Selection implements Representation {
+  id: ID!
+  metadata: RepresentationMetadata!
+  targetObjectId: String!
+  message: String
+  displayedAsTree: Boolean!
+  objects: [SelectionObject!]!
```

Comment:
You need to reuse a tree subscription instead of the raw selection subscription with its list of objects. I don't understand why this needs to be modified and why `displayedAsTree` should exist on both `Selection` and its description.

Reply:
Ok with that but I still have a problem with the "message"

I agree to avoid opening a subscription toward the SelectionRepresentation if the SelectionDialog is intended to represent selectable elements as a tree (in this case we will rely on the TreeView).

The message specified by the specifier is for now not interpreted but the SelectionDescription java pojo code relies on a messageProvider (Function<VariableManager, String>) so it could be interpreted in the future.

So in this case, I'm not sure what I should do to compute the dialog message which is common for both flat and tree representations. 

Should I provide the displayAsTree and the message statically by passing them in the graphQL API :

instead of :
```
type SingleClickOnDiagramElementTool implements Tool {
  id: ID!
  label: String!
  iconURL: [String!]!
  appliesToDiagramRoot: Boolean!
  dialogDescriptionId: String
  targetDescriptions: [DiagramElementDescription!]!
}
```

We will have :

```
type SingleClickOnDiagramElementTool implements Tool {
  id: ID!
  label: String!
  iconURL: [String!]!
  appliesToDiagramRoot: Boolean!
  dialogDescription: DialogDescription
  targetDescriptions: [DiagramElementDescription!]!
}



type DialogDescription {
  id: String!
  parameters; [DialogDescriptionParamter!]!
}

type DialogDescriptionParamter {
  key: String!
  value; String!
}

```